src/main/scala/it/pagopa/interop/agreementprocess/common/readmodel/ReadModelQueries.scala

```diff
@@ -0,0 +1,133 @@
+    limit: Int
+  )(readModel: ReadModelService)(implicit ec: ExecutionContext): Future[PaginatedResult[PersistentAgreement]] = {
+
+    val query: Bson =
+      listAgreementsFilters(eServicesIds, consumersIds, producersIds, descriptorsIds, states, showOnlyUpgradeable)
+
+    val filterPipeline: Seq[Bson] = Seq(`match`(query), lookup("eservices", "data.eserviceId", "data.id", "eservices"))
+
+    val upgradablePipeline: Seq[Bson] = {
+      if (showOnlyUpgradeable) {
+        Seq(
+          unwind(fieldName = "eservices.data.descriptors"),
+          addFields(
+            Field(
+              "currentDescriptor",
+              Document("""{ $filter: {
+              input: "$eservices.data.descriptors",
+              as: "descr",         
+              cond: {  $eq: ["$$descr.id" , "$data.descriptorId"]}}} }""")
+            )
+          ),
+          unwind(fieldName = "$currentDescriptor"),
+          addFields(
+            Field(
+              "upgradableDescriptor",
+              Document(
+                """{ $filter: {
+              input: "$eservices.data.descriptors",
+              as: "upgradable",         
+              cond: { $and: [
+                      {$gt:[ {"$toInt" : "$$upgradable.version"} , {"$toInt": "$currentDescriptor.version"}]}, 
+                      {$in:["$$upgradable.state", [""" + CatalogManagement.Published + """, """ +
+                  CatalogManagement.Suspended + """]]}
+                  ]}}} }"""
+              )
+            )
+          ),
+          `match`(Filters.exists("upgradableDescriptor.0", true))
+        )
+
+      } else
+        Seq.empty
+    }
+
+    for {
+      agreements <- readModel.aggregate[PersistentAgreement](
+        "agreements",
+        filterPipeline ++ upgradablePipeline ++
+          Seq(
+            project(
+              fields(include("data"), computed("lowerName", Document("""{ "$toLower" : "$eservices.data.name" }""")))
+            ),
+            sort(ascending("lowerName"))
+          ),
+        offset = offset,
+        limit = limit
+      )
+      count      <- readModel.aggregate[TotalCountResult](
+        "agreements",
+        Seq(
+          `match`(query),
+          count("totalCount"),
+          project(computed("data", Document("""{ "totalCount" : "$totalCount" }""")))
+        ),
+        offset = 0,
+        limit = Int.MaxValue
+      )
+
+    } yield PaginatedResult(results = agreements, totalCount = count.headOption.map(_.totalCount).getOrElse(0))
+
+  }
+
+  def listAgreementsFilters(
+    eServicesIds: List[String],
+    consumersIds: List[String],
+    producersIds: List[String],
+    descriptorsIds: List[String],
+    states: List[AgreementState],
+    showOnlyUpgradeable: Boolean
+  ): Bson = {
+
+    val statesFilter: Option[Bson] =
+      if (showOnlyUpgradeable)
+        listStatesFilter(List(AgreementState.DRAFT, AgreementState.ACTIVE, AgreementState.SUSPENDED))
+      else
+        listStatesFilter(states)
```

Comment:
if we do something like this we can allow the user to select states even with the `showOnlyUpgradeable` flag enabled
```suggestion
      states match {
        case Nil if showOnlyUpgradeable => upgreadableStates
        case list if showOnlyUpgradeable => upgreadableStates.intersect(list)
        case other => other
      }
```

Reply:
If showOnlyUpgradeable is true , always  we need to overwrite the states passed anyway. See https://pagopa.atlassian.net/browse/PIN-1305

Reply:
not necessarily, a possible use case: 
a user wants to see only upgradable agreements that are currently ACTIVE (but not DRAFT or SUSPENDED).

This case is not excluded from requirements and can be satisfied with the suggested change

Reply:
Ok amend it